server/http/restapi/embedded_form.go

```diff
@@ -0,0 +1,166 @@
+package restapi
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/mattermost/mattermost-plugin-apps/server/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/store"
+	"github.com/mattermost/mattermost-plugin-apps/server/utils/httputils"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	embeddedSubmissionPostIDKey = "mm_postID"
+	embeddedSubmissionAppIDKey  = "mm_appID"
+)
+
+func (a *api) handleEmbeddedForm(w http.ResponseWriter, req *http.Request, userID string) {
+	var dialogRequest model.SubmitDialogRequest
+	err := json.NewDecoder(req.Body).Decode(&dialogRequest)
+	if err != nil {
+		writeDialogError(w, "Could not decode request.")
+		return
+	}
+
+	if dialogRequest.Submission[embeddedSubmissionPostIDKey] == nil {
+		writeDialogError(w, "No Post ID provided.")
+		return
+	}
+
+	postID, ok := dialogRequest.Submission[embeddedSubmissionPostIDKey].(string)
+	if !ok {
+		writeDialogError(w, "Post ID not recognized.")
+		return
+	}
+
+	delete(dialogRequest.Submission, embeddedSubmissionPostIDKey)
+
+	if dialogRequest.Submission[embeddedSubmissionAppIDKey] == nil {
+		writeDialogError(w, "No App ID provided")
+		return
+	}
+
+	appID, ok := dialogRequest.Submission[embeddedSubmissionAppIDKey].(string)
+	if !ok {
+		writeDialogError(w, "App ID not recognized")
+		return
+	}
+
+	delete(dialogRequest.Submission, embeddedSubmissionAppIDKey)
+
+	c := &apps.Call{
+		Wish: store.NewWish(appID, dialogRequest.URL),
+		Request: &apps.CallRequest{
+			Context: &apps.Context{
+				AppID:        store.AppID(appID),
+				ActingUserID: dialogRequest.UserId,
+				ChannelID:    dialogRequest.ChannelId,
+				TeamID:       dialogRequest.TeamId,
+				UserID:       dialogRequest.UserId,
+				PostID:       postID,
+			},
+			Values: apps.FormValues{
+				Data: dialogRequest.Submission,
+			},
+		},
+	}
+
+	resp, err := a.apps.Client.PostWish(c)
+	if err != nil {
+		writeDialogError(w, "Error contacting the app: "+err.Error())
+		return
+	}
+
+	var dialogResponse model.SubmitDialogResponse
+
+	if resp.Type == apps.ResponseTypeError {
+		if resp.Data["errors"] != nil {
+			dialogResponse.Errors = make(map[string]string)
+			if errors, ok := resp.Data["errors"].(map[string]interface{}); ok {
+
+				for key, value := range errors {
+					if svalue, ok := value.(string); ok {
+						dialogResponse.Errors[key] = svalue
+					}
+				}
+			}
+		}
+		dialogResponse.Error = resp.Error
```

Comment:
Do we want to return an error code in the response here, or does the response need to be 200?

Reply:
I believe this is expected to work with a 200 response, not an error response. The thinking behind is "the transaction was succesful, and you have to show an error message".